common/http/src/retry.rs

```diff
@@ -103,95 +106,118 @@ impl Retry {
-                }
-            }
-        ).await.unwrap(/*FIXME handle properly*/)?;
-        let mut file = BufWriter::new(File::from_std(std_file));
-        file.write_all(b",").await?;
-        file
-    } else {
-        file
-    };
-    file.write_all(b"\"body\":").await?;
-    let mut reader = body.reader();
-    let _bytes_written = futures::io::copy(&mut reader, &mut file.compat_mut()).await?;
-    file.write_all(b"}").await?;
-    file.flush().await?;
-
-    Ok(rename(
-        file_name,
-        format!("/tmp/logdna/{}_{}.retry", fn_ts, fn_uuid),
-    )
-    .await?)
-}
 
-pub fn retry_stream(
-    retry_base_delay: Duration,
-) -> impl Stream<Item = Result<(IngestBodyBuffer, Option<Vec<Offset>>), Error>> {
-    let retry = Retry::new(retry_base_delay);
-    stream::unfold(retry, |state| async move {
-        loop {
-            // Try to populate retry queue
-            if state.waiting.is_empty() {
-                if let Err(e) = state.fill_waiting().await {
-                    return Some((Err(e), state));
-                };
-                // If there are still no objects then sleep for a while
+    pub fn into_stream(
+        self,
+    ) -> impl Stream<Item = Result<(IngestBodyBuffer, Option<Vec<Offset>>), Error>> {
+        stream::unfold(self, |state| async move {
+            loop {
+                // Try to populate retry queue
                 if state.waiting.is_empty() {
-                    Delay::new(Duration::from_secs(
-                        state.retry_base_delay_secs.try_into().unwrap_or(2),
-                    ))
-                    .await;
+                    if let Err(e) = state.fill_waiting().await {
+                        return Some((Err(e), state));
+                    };
+                    // If there are still no objects then sleep for a while
+                    if state.waiting.is_empty() {
+                        Delay::new(Duration::from_secs(
+                            state.retry_base_delay_secs.try_into().unwrap_or(2),
+                        ))
+                        .await;
+                    }
                 }
-            }
 
-            if let Some(path) = state.waiting.pop() {
-                match Retry::read_from_disk(&path).await {
-                    Ok((offsets, ingest_body)) => {
-                        match IntoIngestBodyBuffer::into(ingest_body).await {
-                            Ok(body_buffer) => return Some((Ok((body_buffer, offsets)), state)),
-                            Err(e) => return Some((Err(e.into()), state)),
+                if let Some(path) = state.waiting.pop() {
+                    match Retry::read_from_disk(&path).await {
+                        Ok((offsets, ingest_body)) => {
+                            match IntoIngestBodyBuffer::into(ingest_body).await {
+                                Ok(body_buffer) => {
+                                    return Some((Ok((body_buffer, offsets)), state))
```

Comment:
I think having an insignificant delay after each retry read (in the order of 20 to 100ms) will help prioritize the current logs.

Reply:
Do you think we should reintroduce the step delay config option?

Reply:
I think the previous delay was in the order of seconds, I'm in favor for something in the order of millis (non-configurable).
Having few tens milliseconds delay will cause the tailed logs to win over this and to send in a lower rate (trying to avoid thundering herd when the service come back online)